README.md
ember-router-service-refresh-polyfill
==============================================================================

Polyfills the RouterService#refresh method as described in [RFC 631](https://emberjs.github.io/rfcs/0631-refresh-method-for-router-service.html).


Compatibility
------------------------------------------------------------------------------

* Ember.js v3.4 or above. The polyfill is disabled in Ember 4.1 and above.
* Ember CLI v3.4 or above
* Embroider safe and optimized
* Node.js v12 or above


Installation
------------------------------------------------------------------------------

```
ember install ember-router-service-refresh-polyfill
```


Usage
------------------------------------------------------------------------------

This addon polyfills the RouterService#refresh method which refreshes all currently active routes, doing a full transition.

If a `routeName` is provided and refers to a currently active route,
it will refresh only that route and its descendents.

It returns a [Transition](https://api.emberjs.com/ember/release/classes/Transition) that will be resolved once the refresh is complete.
All resetController, beforeModel, model, afterModel, redirect, and setupController
hooks will be called again. You will get new data from the model hook.

### Example

```js
import { action } from '@ember/object';
import { inject as service } from '@ember/service';
import Component from '@glimmer/component';

export default class VeryRealComponent extends Component {
  @service('router') routerService;

  @action
  refreshAllActiveRoutes() {
    this.routerService.refresh();
  }

  @action
  refreshNamedActiveRoute() {
    this.routerService.refresh('some-active-route-name');
  }
}
```

Contributing
------------------------------------------------------------------------------

See the [Contributing](CONTRIBUTING.md) guide for details.


License
------------------------------------------------------------------------------

This project is licensed under the [MIT License](LICENSE.md).


CHANGELOG.md
# Changelog



## v1.1.0 (2022-11-15)

#### :rocket: Enhancement
* [#16](https://github.com/ember-polyfills/ember-router-service-refresh-polyfill/pull/16) Convert ember-cli-htmlbars to a devDependency ([@Windvis](https://github.com/Windvis))

#### :house: Internal
* [#19](https://github.com/ember-polyfills/ember-router-service-refresh-polyfill/pull/19) Use Github Actions to publish to npm ([@Windvis](https://github.com/Windvis))
* [#17](https://github.com/ember-polyfills/ember-router-service-refresh-polyfill/pull/17) Update to the v4.3.0 blueprints ([@Windvis](https://github.com/Windvis))

#### Committers: 1
- Sam Van Campenhout ([@Windvis](https://github.com/Windvis))

## v1.0.1 (2022-01-19)

#### :house: Internal
* [#15](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/15) Update dependencies ([@Windvis](https://github.com/Windvis))

#### Committers: 1
- Sam Van Campenhout ([@Windvis](https://github.com/Windvis))


## v1.0.0 (2021-11-24)

#### :boom: Breaking Change
* [#13](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/13) Drop support for node 10 ([@Windvis](https://github.com/Windvis))

#### :rocket: Enhancement
* [#10](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/10) Disable the polyfill when the method is available in ember-source ([@Windvis](https://github.com/Windvis))

#### :house: Internal
* [#12](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/12) Update to the Ember 3.28 addon blueprint ([@Windvis](https://github.com/Windvis))
* [#14](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/14) Update to ember-auto-import v2 ([@Windvis](https://github.com/Windvis))

#### Committers: 1
- Sam Van Campenhout ([@Windvis](https://github.com/Windvis))


## v0.1.0 (2021-03-20)

#### :rocket: Enhancement
* [#4](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/4) Add Ember 3.4 support ([@Windvis](https://github.com/Windvis))
* [#3](https://github.com/Windvis/ember-router-service-refresh-polyfill/pull/3) Add a refresh method to the RouterService ([@Windvis](https://github.com/Windvis))

#### Committers: 1
- Sam Van Campenhout ([@Windvis](https://github.com/Windvis))


tests/dummy/app/router.js
/* eslint-disable ember/no-classic-classes */
import EmberRouter from '@ember/routing/router';
import config from './config/environment';

const Router = EmberRouter.extend({
  location: config.locationType,
  rootURL: config.rootURL,
});

Router.map(function () {
  this.route('parent', { path: '/' }, function () {
    this.route('child');
    this.route('sister');
  });
});

export default Router;


tests/helpers/should-test-polyfill.js
import { test, skip } from 'qunit';
import { SHOULD_POLYFILL_ROUTER_SERVICE_REFRESH } from 'ember-router-service-refresh-polyfill/initializers/setup-router-service-refresh-polyfill';

export const testIfPolyfilled = SHOULD_POLYFILL_ROUTER_SERVICE_REFRESH
  ? test
  : skip;


tests/acceptance/router-service-refresh-test.js
/* eslint-disable ember/no-classic-classes */

import Route from '@ember/routing/route';
import { visit } from '@ember/test-helpers';
import { setupApplicationTest } from 'ember-qunit';
import { module, test } from 'qunit';

module('Acceptance | RouterService | refresh', function (hooks) {
  setupApplicationTest(hooks);

  test('it can be used to re-run the model hooks of active routes', async function (assert) {
    const routerService = this.owner.lookup('service:router');

    let parentCounter = 0;
    this.owner.register(
      'route:parent',
      Route.extend({
        model() {
          ++parentCounter;
        },
      })
    );

    let childCounter = 0;
    this.owner.register(
      'route:parent.child',
      Route.extend({
        model() {
          ++childCounter;
        },
      })
    );

    let sisterCounter = 0;
    this.owner.register(
      'route:parent.sister',
      Route.extend({
        model() {
          ++sisterCounter;
        },
      })
    );

    await visit('/');
    assert.strictEqual(parentCounter, 1);
    assert.strictEqual(childCounter, 0);
    assert.strictEqual(sisterCounter, 0);

    await routerService.refresh();
    assert.strictEqual(parentCounter, 2);
    assert.strictEqual(childCounter, 0);
    assert.strictEqual(sisterCounter, 0);

    await routerService.refresh('application');
    assert.strictEqual(parentCounter, 3);
    assert.strictEqual(childCounter, 0);
    assert.strictEqual(sisterCounter, 0);

    await routerService.transitionTo('parent.child');
    assert.strictEqual(parentCounter, 3);
    assert.strictEqual(childCounter, 1);
    assert.strictEqual(sisterCounter, 0);

    await routerService.refresh('parent.child');
    assert.strictEqual(parentCounter, 3);
    assert.strictEqual(childCounter, 2);
    assert.strictEqual(sisterCounter, 0);

    await routerService.refresh('parent');
    assert.strictEqual(parentCounter, 4);
    assert.strictEqual(childCounter, 3);
    assert.strictEqual(sisterCounter, 0);

    await routerService.transitionTo('parent.sister');
    assert.strictEqual(parentCounter, 4);
    assert.strictEqual(childCounter, 3);
    assert.strictEqual(sisterCounter, 1);

    await routerService.refresh();
    assert.strictEqual(parentCounter, 5);
    assert.strictEqual(childCounter, 3);
    assert.strictEqual(sisterCounter, 2);
  });

  test('it verifies that the provided route exists', async function (assert) {
    const routerService = this.owner.lookup('service:router');

    assert.throws(() => {
      routerService.refresh('this-route-does-not-exist');
    });
  });

  test('it verifies that the provided route is active', async function (assert) {
    const routerService = this.owner.lookup('service:router');

    assert.throws(() => {
      routerService.refresh('this-route-does-not-exist');
    });
  });
});


addon/initializers/setup-router-service-refresh-polyfill.js
/* eslint-disable ember/no-private-routing-service */
import { assert } from '@ember/debug';
import { getOwner } from '@ember/application';
import { dependencySatisfies, macroCondition } from '@embroider/macros';

export const SHOULD_POLYFILL_ROUTER_SERVICE_REFRESH = !dependencySatisfies(
  'ember-source',
  '^4.1.0-beta.1' // https://github.com/emberjs/ember.js/blob/master/CHANGELOG.md#v410-beta1-november-19-2021
);

export function initialize(application) {
  if (macroCondition(SHOULD_POLYFILL_ROUTER_SERVICE_REFRESH)) {
    const RouterService = application.resolveRegistration('service:router');

    RouterService.reopen({
      /**
       * Refreshes all currently active routes, doing a full transition.
       * If a routeName is provided and refers to a currently active route,
       * it will refresh only that route and its descendents.
       * Returns a promise that will be resolved once the refresh is complete.
       * All resetController, beforeModel, model, afterModel, redirect, and setupController
       * hooks will be called again. You will get new data from the model hook.
       *
       * @method refresh
       * @param {String} [routeName] the route to refresh (along with all child routes)
       * @return Transition
       * @public
       */
      refresh(routeName) {
        if (!routeName) {
          return this._router._routerMicrolib.refresh();
        }

        assert(
          `The route "${routeName}" was not found`,
          this._router.hasRoute(routeName)
        );

        assert(
          `The route "${routeName}" is currently not active`,
          this.isActive(routeName)
        );

        let owner = getOwner(this);
        let pivotRoute = owner.lookup(`route:${routeName}`);

        return this._router._routerMicrolib.refresh(pivotRoute);
      },
    });
  }
}

export default {
  initialize,
};


config/ember-try.js
'use strict';

const getChannelURL = require('ember-source-channel-url');
const { embroiderSafe, embroiderOptimized } = require('@embroider/test-setup');

module.exports = async function () {
  return {
    useYarn: true,
    scenarios: [
      {
        name: 'ember-lts-3.4',
        npm: {
          devDependencies: {
            'ember-source': '~3.4.0',
          },
        },
      },
      {
        name: 'ember-lts-3.24',
        npm: {
          devDependencies: {
            'ember-source': '~3.24.3',
          },
        },
      },
      {
        name: 'ember-lts-3.28',
        npm: {
          devDependencies: {
            'ember-source': '~3.28.0',
          },
        },
      },
      {
        name: 'ember-4.0',
        npm: {
          devDependencies: {
            'ember-source': '~4.0.0',
          },
        },
      },
      {
        name: 'ember-release',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('release'),
          },
        },
      },
      {
        name: 'ember-beta',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('beta'),
          },
        },
      },
      {
        name: 'ember-canary',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('canary'),
          },
        },
      },
      {
        name: 'ember-classic',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({
            'application-template-wrapper': true,
            'default-async-observers': false,
            'template-only-glimmer-components': false,
          }),
        },
        npm: {
          devDependencies: {
            'ember-source': '~3.28.0',
          },
          ember: {
            edition: 'classic',
          },
        },
      },
      embroiderSafe(),
      embroiderOptimized(),
    ],
  };
};
